workers/unity/Packages/io.improbable.gdk.core/Systems/EntitySystem.cs

```diff
@@ -1,61 +1,127 @@
             {
-                entitiesAdded.Clear();
-                entitiesRemoved.Clear();
+                EntityIds = new NativeArray<EntityId>(size, Allocator.Persistent);
+                EntityCount = 0;
+                JobHandle = default;
+            }
 
-                // todo decide on a container and remove this
-                foreach (var entityId in diff.GetEntitiesAdded())
+            public NativeSlice<EntityId> Slice
+            {
+                get
                 {
-                    entitiesAdded.Add(entityId);
+                    JobHandle.Complete();
+                    return new NativeSlice<EntityId>(EntityIds, 0, EntityCount);
                 }
+            }
 
-                foreach (var entityId in diff.GetEntitiesRemoved())
+            public void EnsureSize(ref EntityQuery query)
+            {
+                EntityCount = query.CalculateEntityCount();
+                if (EntityIds.Length >= EntityCount)
                 {
-                    entitiesRemoved.Add(entityId);
+                    return;
                 }
 
-                if (entitiesAdded.Count != 0 || entitiesRemoved.Count != 0)
+                EntityIds.Dispose();
+                EntityIds = new NativeArray<EntityId>(EntityCount << 1, Allocator.Persistent);
+            }
+
+            public void Dispose()
+            {
+                if (EntityIds.IsCreated)
                 {
-                    ViewVersion += 1;
+                    EntityIds.Dispose();
                 }
             }
         }
 
+        private EndSimulationEntityCommandBufferSystem bufferSystem;
+        private EntityCollection added;
+        private EntityQuery addedQuery; // This query needs to be a member of SystemBase
+        private EntityCollection removed;
+        private EntityQuery removedQuery; // This query needs to be a member of SystemBase
+        public int ViewVersion { get; private set; }
+        public NativeSlice<EntityId> EntitiesRemoved => removed.Slice;
+        public NativeSlice<EntityId> EntitiesAdded => added.Slice;
+
+        protected override void OnCreate()
+        {
+            base.OnCreate();
+            bufferSystem = World.GetOrCreateSystem<EndSimulationEntityCommandBufferSystem>();
+            removed = new EntityCollection(1);
+            added = new EntityCollection(1);
+        }
+
         protected override void OnUpdate()
         {
-            entitiesAdded.Clear();
-            entitiesRemoved.Clear();
+            UpdateEntitiesAdded();
+            UpdateEntitiesRemoved();
+
+            Dependency = JobHandle.CombineDependencies(added.JobHandle, removed.JobHandle);
+            bufferSystem.AddJobHandleForProducer(Dependency);
+
+            if (added.EntityCount != 0 || removed.EntityCount != 0)
+            {
+                ViewVersion += 1;
+            }
+        }
+
+        private void UpdateEntitiesAdded()
+        {
+            added.EnsureSize(ref addedQuery);
```

Comment:
This will only ensure the size is right for the amount of entities we had in the PREVIOUS iteration, that doesn't sound right?

Reply:
Apparently it doesn't, I got the example from [Unity's docs](https://docs.unity3d.com/Packages/com.unity.entities@0.2/manual/entities_job_foreach.html) (in the 'Access to EntityQuery object example' section) and the unit test doesn't fail.